[Ripple] Improve MDCRippleTouchController

maicki opened the pull request:
Changes:
- Create new initializer that allows to defer adding `MDCRippleView` to the a provided view
- Add `MDCRippleView` on first touch down if YES was passed in as deferred parameter
- Lazy getter for `rippleView`
- As `rippleView` property is not supposed to be overwritten always use `_rippleView` instance variable within touch controller
- Cache all delegate methods for faster lookup if a delegate responds to a selector

Determined improvements:
- Reduce memory overhead and safes a couple of cycles due to creation of `MDCRippleView` lazily
- Reduce overhead for adding / removing `MDCRippleView` to view hierarchies
- Reduce overhead of having the `MDCRippleView` that most of the time is hidden within view hierarchies as it will only get's added and immediately removed if needed

maicki committed:
Improve MDCRippleTouchController

Changes:
- Create MDCRippleView lazily
- Add MDCRippleView in gesture recognizer callback
- Remove MDCRippleView after ripple animation
- As rippleView property is not supposed to be overwritten always use _rippleView instance variable within touch controller

Determined improvements:
- Reduce memory overhead and safes a couple of cycles due to creation of MDCRippleView lazily
- Reduce overhead for adding / removing MDCRippleView to view hierarchies
- Reduce overhead of having the MDCRippleView that most of the time is hidden within view hierarchies as it will only get's added and immediately removed if needed

maicki committed:
Add new initializer for adding the ripple view deferred

yarneo reviewed (approved):
Thank you!

material-automation commented on components/Ripple/src/MDCRippleTouchController.m:
```diff
@@ -38,70 +45,98 @@ - (instancetype)init {
+      [_delegate respondsToSelector:@selector(rippleTouchController:
+                                               didProcessRippleView:atTouchLocation:)];
+  _delegateFlags.rippleTouchControllerInsertRippleViewIntoView = [delegate
+      respondsToSelector:@selector(rippleTouchController:insertRippleView:intoView:)];
```
clang-format suggested the following change:


```
-  _delegateFlags.rippleTouchControllerDidProcessRippleViewAtTouchLocation =
-      [_delegate respondsToSelector:@selector(rippleTouchController:
-                                               didProcessRippleView:atTouchLocation:)];
-  _delegateFlags.rippleTouchControllerInsertRippleViewIntoView = [delegate
-      respondsToSelector:@selector(rippleTouchController:insertRippleView:intoView:)];
+  _delegateFlags.rippleTouchControllerDidProcessRippleViewAtTouchLocation = [_delegate
+      respondsToSelector:@selector(rippleTouchController:didProcessRippleView:atTouchLocation:)];
+  _delegateFlags.rippleTouchControllerInsertRippleViewIntoView =
+      [delegate respondsToSelector:@selector(rippleTouchController:insertRippleView:intoView:)];
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/src/MDCRippleTouchController.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented:
bazel detected changes to the following targets:

```
//components/ActionSheet:snapshot_tests
//components/ActionSheet:snapshot_tests_test_bundle
//components/ActionSheet:unit_tests
//components/ActionSheet:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/ActionSheet:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/ActionSheet:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/ActionSheet:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/ActionSheet:unit_tests_IPHONE_X_IN_11_0
//components/ActionSheet:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/AppBar:snapshot_tests
//components/AppBar:snapshot_tests_test_bundle
//components/AppBar:unit_tests
//components/AppBar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/AppBar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/AppBar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/AppBar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/AppBar:unit_tests_environment
//components/AppBar:unit_tests_environment_test_bundle
//components/Banner:snapshot_tests
//components/Banner:snapshot_tests_test_bundle
//components/Banner:unit_tests
//components/Banner:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Banner:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Banner:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Banner:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Banner:unit_tests_environment
//components/Banner:unit_tests_environment_test_bundle
//components/BottomAppBar:snapshot_tests
//components/BottomAppBar:snapshot_tests_test_bundle
//components/BottomAppBar:unit_tests
//components/BottomAppBar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/BottomAppBar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/BottomAppBar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/BottomAppBar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/BottomAppBar:unit_tests_environment
//components/BottomAppBar:unit_tests_environment_test_bundle
//components/BottomNavigation:snapshot_tests
//components/BottomNavigation:snapshot_tests_test_bundle
//components/BottomNavigation:unit_tests
//components/BottomNavigation:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/BottomNavigation:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/BottomNavigation:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/BottomNavigation:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/BottomNavigation:unit_tests_beta
//components/BottomNavigation:unit_tests_beta_test_bundle
//components/BottomNavigation:unit_tests_environment
//components/BottomNavigation:unit_tests_environment_test_bundle
//components/ButtonBar:snapshot_tests
//components/ButtonBar:snapshot_tests_test_bundle
//components/ButtonBar:unit_tests
//components/ButtonBar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/ButtonBar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/ButtonBar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/ButtonBar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/ButtonBar:unit_tests_IPHONE_X_IN_11_0
//components/ButtonBar:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/Buttons:snapshot_tests
//components/Buttons:snapshot_tests_test_bundle
//components/Buttons:unit_tests
//components/Buttons:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Buttons:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Buttons:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Buttons:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Buttons:unit_tests_environment
//components/Buttons:unit_tests_environment_test_bundle
//components/Cards:snapshot_tests
//components/Cards:snapshot_tests_test_bundle
//components/Cards:unit_tests
//components/Cards:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Cards:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Cards:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Cards:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Cards:unit_tests_environment
//components/Cards:unit_tests_environment_test_bundle
//components/Chips:snapshot_tests
//components/Chips:snapshot_tests_test_bundle
//components/Chips:unit_tests
//components/Chips:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Chips:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Chips:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Chips:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Chips:unit_tests_environment
//components/Chips:unit_tests_environment_test_bundle
//components/Dialogs:snapshot_tests
//components/Dialogs:snapshot_tests_test_bundle
//components/Dialogs:unit_tests
//components/Dialogs:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Dialogs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Dialogs:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Dialogs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Dialogs:unit_tests_environment
//components/Dialogs:unit_tests_environment_test_bundle
//components/List:snapshot_tests
//components/List:snapshot_tests_test_bundle
//components/List:unit_tests
//components/List:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/List:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/List:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/List:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/List:unit_tests_environment
//components/List:unit_tests_environment_test_bundle
//components/NavigationBar:snapshot_tests
//components/NavigationBar:snapshot_tests_test_bundle
//components/NavigationBar:unit_tests
//components/NavigationBar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/NavigationBar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/NavigationBar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/NavigationBar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/NavigationBar:unit_tests_IPHONE_X_IN_11_0
//components/NavigationBar:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/Ripple:snapshot_tests
//components/Ripple:snapshot_tests_test_bundle
//components/Ripple:unit_tests
//components/Ripple:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Ripple:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Ripple:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Ripple:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Ripple:unit_tests_environment
//components/Ripple:unit_tests_environment_test_bundle
//components/ShadowElevations:snapshot_tests
//components/ShadowElevations:snapshot_tests_test_bundle
//components/Slider:snapshot_tests
//components/Slider:snapshot_tests_test_bundle
//components/Slider:unit_tests
//components/Slider:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Slider:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Slider:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Slider:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Slider:unit_tests_environment
//components/Slider:unit_tests_environment_test_bundle
//components/Snackbar:snapshot_tests
//components/Snackbar:snapshot_tests_test_bundle
//components/Snackbar:unit_tests
//components/Snackbar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Snackbar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Snackbar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Snackbar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Snackbar:unit_tests_environment
//components/Snackbar:unit_tests_environment_test_bundle
//components/Tabs:snapshot_tests
//components/Tabs:snapshot_tests_test_bundle
//components/Tabs:unit_tests
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Tabs:unit_tests_environment
//components/Tabs:unit_tests_environment_test_bundle
//components/private/ThumbTrack:unit_tests
//components/private/ThumbTrack:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/private/ThumbTrack:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/private/ThumbTrack:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/private/ThumbTrack:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/private/ThumbTrack:unit_tests_environment
//components/private/ThumbTrack:unit_tests_environment_test_bundle
```

<!-- identifier: affected-targets -->

jverkoey reviewed (approved):
Wonderful, thank you!

yarneo commented:
I believe there are some tests that rely on the bounds to be correct and the rippleView to be added as a subview earlier in the lifecycle.

Specifically these are failing:
```
ActionSheetRippleTests testRippleIsEnabledAndInkIsDisabledForAllCellsAndTheirPropertiesAreCorrectWhenRippleBehaviorIsEnabled
MDCTabBarRippleTests testEnabledRippleAndDisabledInkPropertiesAndBehaviorWhenEnabledRippleBehavior
BottomNavigationRippleTests testEnabledInkAndDisabledRippleColorsAndSuperviewsAndBounds
BottomNavigationRippleTests testEnabledRippleAndDisabledInkColorsAndSuperviewsAndBoundsWithRippleBehaviorEnabled
MDCRippleTouchControllerTests testAddRippleToView
MDCRippleTouchControllerTests testCallsInsertRippleViewDelegateMethodWhenAddedToView
MDCRippleTouchControllerTests testInitWithView
MDCTabBarRippleTests testEnabledRippleAndDisabledInkPropertiesAndBehaviorWhenEnabledRippleBehavior
also a couple of MDCTabBarViewItemViewSnapshotTests snapshot tests
```

material-automation commented on components/Ripple/tests/unit/MDCRippleTouchControllerTests.m:
```diff
@@ -17,6 +17,7 @@
 #import <UIKit/UIGestureRecognizerSubclass.h>
 
 #import "MaterialRipple.h"
```
clang-format suggested the following change:


```
-#import "MaterialRipple.h"
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/tests/unit/MDCRippleTouchControllerTests.m -->

material-automation replied:
clang-format suggested the following change:


```
+#import "MaterialRipple.h"
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/tests/unit/MDCRippleTouchControllerTests.m -->

material-automation commented on components/Tabs/tests/unit/MDCTabBarRippleTests.m:
```diff
@@ -20,6 +20,7 @@
 #import "MaterialInk.h"
 #import "MaterialRipple.h"
 #import "MaterialTabs.h"
+#import "MDCRippleTouchController+Testing.h"
```
clang-format suggested the following change:


```
-#import "MDCRippleTouchController+Testing.h"
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/unit/MDCTabBarRippleTests.m -->

material-automation commented on components/ActionSheet/tests/unit/ActionSheetRippleTests.m:
```diff
@@ -19,6 +19,7 @@
 #import "MaterialActionSheet.h"
 #import "MaterialInk.h"
 #import "MaterialRipple.h"
+#import "MDCRippleTouchController+Testing.h"
```
clang-format suggested the following change:


```
-#import "MDCRippleTouchController+Testing.h"
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/unit/ActionSheetRippleTests.m -->

material-automation commented on components/BottomNavigation/tests/unit/BottomNavigationRippleTests.m:
```diff
@@ -18,6 +18,7 @@
 #import "MaterialBottomNavigation.h"
 #import "MaterialInk.h"
 #import "MaterialRipple.h"
+#import "MDCRippleTouchController+Testing.h"
```
clang-format suggested the following change:


```
-#import "MDCRippleTouchController+Testing.h"
```

<!-- identifier: clang-format -->
<!-- File: components/BottomNavigation/tests/unit/BottomNavigationRippleTests.m -->

material-automation commented on components/BottomNavigation/tests/unit/BottomNavigationRippleTests.m:
```diff
@@ -90,7 +91,12 @@ - (void)testEnabledRippleAndDisabledInkColorsAndSuperviewsAndBoundsWithRippleBeh
+
+    XCTAssertNil(itemView.rippleTouchController.rippleView.superview);
+    itemView.rippleTouchController.gestureRecognizer.state = UIGestureRecognizerStateBegan;
+    [itemView.rippleTouchController handleRippleGesture:itemView.rippleTouchController.gestureRecognizer];
```
clang-format suggested the following change:


```
-    [itemView.rippleTouchController handleRippleGesture:itemView.rippleTouchController.gestureRecognizer];
+    [itemView.rippleTouchController
+        handleRippleGesture:itemView.rippleTouchController.gestureRecognizer];
```

<!-- identifier: clang-format -->
<!-- File: components/BottomNavigation/tests/unit/BottomNavigationRippleTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

yarneo commented:
To resolve the bazel build error you will need to:

add to the Ripple/BUILD file visibility to the "private" target to the other test targets like so:

```
mdc_objc_library(
    name = "private",
    hdrs = native.glob(["src/private/*.h"]),
    includes = ["src/private"],
    visibility = [":test_targets"],
)

package_group(
    name = "test_targets",
    packages = [
        "//components/ActionSheet/...",
        "//components/BottomNavigation/...",
        "//components/Tabs/...",
    ],
)
```

You will then have to go to the unit_test target declarations in those component BUILD files and add:
```
"//components/Ripple:private",
```

As a dependency

maicki commented:
@yarneo Yeah already discovered that this afternoon. Thanks for pointing out though, greatly appreciated!

material-automation commented on components/Tabs/tests/unit/MDCTabBarRippleTests.m:
```diff
@@ -19,6 +19,7 @@
 #import "../../src/private/MDCItemBarStyle.h"
 #import "MaterialInk.h"
 #import "MaterialRipple.h"
+#import "../../../Ripple/src/private/MDCRippleTouchController+Testing.h"
```
clang-format suggested the following change:


```
-#import "../../../Ripple/src/private/MDCRippleTouchController+Testing.h"
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/unit/MDCTabBarRippleTests.m -->

material-automation commented on components/ActionSheet/tests/unit/ActionSheetRippleTests.m:
```diff
@@ -19,6 +19,7 @@
 #import "MaterialActionSheet.h"
 #import "MaterialInk.h"
 #import "MaterialRipple.h"
+#import "../../../Ripple/src/private/MDCRippleTouchController+Testing.h"
```
clang-format suggested the following change:


```
-#import "../../../Ripple/src/private/MDCRippleTouchController+Testing.h"
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/unit/ActionSheetRippleTests.m -->

material-automation commented on components/BottomNavigation/tests/unit/BottomNavigationRippleTests.m:
```diff
@@ -18,6 +18,7 @@
 #import "MaterialBottomNavigation.h"
 #import "MaterialInk.h"
 #import "MaterialRipple.h"
+#import "../../../Ripple/src/private/MDCRippleTouchController+Testing.h"
```
clang-format suggested the following change:


```
-#import "../../../Ripple/src/private/MDCRippleTouchController+Testing.h"
```

<!-- identifier: clang-format -->
<!-- File: components/BottomNavigation/tests/unit/BottomNavigationRippleTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

romoore commented on components/Typography/tests/unit/UIFont+MaterialTypographyTests.m:
```diff
@@ -69,7 +69,7 @@ - (void)testScaledFontForTraitEnvironmentWithNoApplicationOniOS9DoesNothing {
 
   // Given
-  XCTAssertNil(UIApplication.sharedApplication);
+  // XCTAssertNil(UIApplication.sharedApplication);
```
It looks like this was accidentally left commented-out, since Typography isn't affected by Ripple. Please uncomment this line.

romoore reviewed (dismissed):
Thanks for making performance improvements to Ripple!

This appears to have broken a number of snapshot tests. The ripples don't appear to be rendering in the snapshots any longer.  This makes me a little concerned that Ripple might not be showing as expected in all cases, but it may also be an artifact of the test set-up. 

|Before|After|
|---|---|
|<img width="1178" alt="reference_testFixedClusteredLeadingLayoutStyleTooWideSizeLatinLTR_11_2@2x" src="https://user-images.githubusercontent.com/1753199/67927902-2409df80-fb90-11e9-8de2-c9da482bbf4b.png">|<img width="1178" alt="failed_testFixedClusteredLeadingLayoutStyleTooWideSizeLatinLTR_11_2@2x" src="https://user-images.githubusercontent.com/1753199/67927977-4f8cca00-fb90-11e9-87f7-69f7be7dfbc5.png">|

This  PR also looks like it could be split into a few different improvements. Would it make sense to land the non-breaking changes first in separate PRs and worry about the potential side-effects of the lazy Ripple (or whatever caused the snapshots to change) separately?

## List of tests failing for me locally

                -[MDCBannerViewSnapshotTests testDynamicTypeForAttributedTextStringWhenContentSizeCategoryIsExtraExtraLarge]
                -[MDCBannerViewSnapshotTests testDynamicTypeForContentSizeCategoryExtraExtraLarge]
                -[MDCBannerViewSnapshotTests testLongTextWithSingleActionLTR]
                -[MDCBannerViewSnapshotTests testLongTextWithSingleActionRTLInArabic]
                -[MDCBannerViewSnapshotTests testLongTextWithTwoActionsAndIconLTR]
                -[MDCBannerViewSnapshotTests testLongTextWithTwoActionsAndIconRTLInArabic]
                -[MDCBannerViewSnapshotTests testLongTextWithTwoActionsLTR]
                -[MDCBannerViewSnapshotTests testLongTextWithTwoActionsRTLInArabic]
                -[MDCBannerViewSnapshotTests testMiddleLengthTextAndStackedButtonsWithLargeLayoutMargin]
                -[MDCBannerViewSnapshotTests testMiddleLengthTextWithLargeLayoutMargin]
                -[MDCBannerViewSnapshotTests testShortTextWithSingleActionLTR]
                -[MDCBannerViewSnapshotTests testShortTextWithSingleActionRTLInArabic]
                -[MDCBannerViewSnapshotTests testSingleRowStyleLongTextWithSingleActionLTR]
                -[MDCBannerViewSnapshotTests testSingleRowStyleLongTextWithSingleActionRTLInArabic]

                -[MDCTabBarViewItemViewSnapshotTests testRippleAppearanceWhenFullyPressed]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredCenteredLayoutContentPaddingArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredCenteredLayoutContentPaddingLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredCenteredLayoutStyleFitSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredCenteredLayoutStyleFitSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredCenteredLayoutStyleTooNarrowSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredCenteredLayoutStyleTooNarrowSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredCenteredLayoutStyleTooWideSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredCenteredLayoutStyleTooWideSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredLeadingLayoutContentPaddingArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredLeadingLayoutContentPaddingLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredLeadingLayoutStyleFitSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredLeadingLayoutStyleFitSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredLeadingLayoutStyleTooNarrowSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredLeadingLayoutStyleTooNarrowSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredLeadingLayoutStyleTooWideSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredLeadingLayoutStyleTooWideSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredTrailingLayoutContentPaddingArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredTrailingLayoutContentPaddingLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredTrailingLayoutStyleFitSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredTrailingLayoutStyleFitSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredTrailingLayoutStyleTooNarrowSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredTrailingLayoutStyleTooNarrowSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredTrailingLayoutStyleTooWideSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedClusteredTrailingLayoutStyleTooWideSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedLayoutContentPaddingArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedLayoutContentPaddingLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedLayoutStyleFitSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedLayoutStyleFitSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedLayoutStyleTooNarrowSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedLayoutStyleTooNarrowSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedLayoutStyleTooWideSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testFixedLayoutStyleTooWideSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testScrollableLayoutContentPaddingArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testScrollableLayoutContentPaddingLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testScrollableLayoutStyleFitSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testScrollableLayoutStyleFitSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testScrollableLayoutStyleTooNarrowSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testScrollableLayoutStyleTooNarrowSizeLatinLTR]
                -[MDCTabBarViewLayoutStyleSnapshotTests testScrollableLayoutStyleTooWideSizeArabicRTL]
                -[MDCTabBarViewLayoutStyleSnapshotTests testScrollableLayoutStyleTooWideSizeLatinLTR]
                -[MDCTabBarViewSnapshotTests testRippleColor]

maicki committed:
Update comment for MDCRippleTouchController and add tests for deferred adding of rippleView

maicki committed:
Remove not needed import

material-automation commented on components/Ripple/tests/unit/MDCRippleTouchControllerTests.m:
```diff
@@ -98,6 +94,19 @@ - (void)testInitWithView {
+  // Given
+  UIView *parentView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 100, 100)];
+  MDCRippleTouchController *touchController =
+  [[MDCRippleTouchController alloc] initWithView:parentView deferred:YES];
```
clang-format suggested the following change:


```
-  [[MDCRippleTouchController alloc] initWithView:parentView deferred:YES];
+      [[MDCRippleTouchController alloc] initWithView:parentView deferred:YES];
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/tests/unit/MDCRippleTouchControllerTests.m -->

material-automation commented on components/Ripple/tests/unit/MDCRippleTouchControllerTests.m:
```diff
@@ -117,6 +126,25 @@ - (void)testAddRippleToView {
+  // Given
+  UIView *parentView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 100, 100)];
+  MDCRippleTouchController *touchController =
+  [[MDCRippleTouchController alloc] initWithView:parentView deferred:YES];
```
clang-format suggested the following change:


```
-  [[MDCRippleTouchController alloc] initWithView:parentView deferred:YES];
+      [[MDCRippleTouchController alloc] initWithView:parentView deferred:YES];
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/tests/unit/MDCRippleTouchControllerTests.m -->

material-automation commented on components/Ripple/tests/unit/MDCRippleTouchControllerTests.m:
```diff
@@ -185,4 +213,5 @@ - (void)testCallsShouldProcessRippleViewDelegateMethodWhenProcessRippleIsDisable
   XCTAssertFalse(delegate.didProcessRippleViewCalled);
 }
 
+
```
clang-format suggested the following change:


```
-
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/tests/unit/MDCRippleTouchControllerTests.m -->

material-automation commented on components/Ripple/src/MDCRippleTouchController.h:
```diff
@@ -76,6 +76,24 @@
+ view if the first on tap event is happening. In case -[MDCRippleTouchControllerDelegate
+ rippleTouchController:insertRippleView:intoView:] *is* implemented, it's the
+ responsibility of the delegate to add the ripple view in the proper position within view's
+ hierarchy if the delegate method is called.
```
clang-format suggested the following change:


```
- Initializes the controller and based on the deferred parameter is adding the ripple view immediately as subview of the priovided view.
-
- Note: If YES is passed for the deferred parameter and the -[MDCRippleTouchControllerDelegate  rippleTouchController:insertRippleView:intoView:] is
- *not* implemented the rippleview will be automatically added as top subview to the given
- view if the first on tap event is happening. In case -[MDCRippleTouchControllerDelegate
- rippleTouchController:insertRippleView:intoView:] *is* implemented, it's the
- responsibility of the delegate to add the ripple view in the proper position within view's
- hierarchy if the delegate method is called.
+ Initializes the controller and based on the deferred parameter is adding the ripple view
+ immediately as subview of the priovided view.
+
+ Note: If YES is passed for the deferred parameter and the -[MDCRippleTouchControllerDelegate
+ rippleTouchController:insertRippleView:intoView:] is *not* implemented the rippleview will be
+ automatically added as top subview to the given view if the first on tap event is happening. In
+ case -[MDCRippleTouchControllerDelegate rippleTouchController:insertRippleView:intoView:] *is*
+ implemented, it's the responsibility of the delegate to add the ripple view in the proper position
+ within view's hierarchy if the delegate method is called.
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/src/MDCRippleTouchController.h -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented:
## API diff detected the following changes

### Ripple

#### MDCRippleTouchController

*new* method: `-initWithView:deferred:` in `MDCRippleTouchController`

<!-- identifier: apidiff -->

maicki committed:
Clang-format

maicki committed:
Use view instead of _view

maicki commented:
@yarneo @jverkoey @romoore  would you be so kind and give it another look. Thank you!

I reverted the changes that would add the `rippleView` deferred by default as this broke the assumption of the current API that the `rippleView` is immediately added in certain cases, what also caused the test breakages.
For now let's go with a new initializer (`initWithView:deferred`) that has another parameter where consumer can decide if the ripple view should be added in a deferred manner or not.

All regular tests passed in the last CI run again as well as I added new tests for the new initializer.

romoore commented on components/Ripple/src/MDCRippleTouchController.h:
```diff
@@ -76,6 +76,25 @@
  */
 - (nonnull instancetype)init;
 
+/**
```
Nit: Doc clean-up:

```objc
/**
 Initializes the controller and based on the deferred parameter is adding the ripple view
 immediately as subview of the priovided view.

 @Note If YES is passed for the deferred parameter and
 @c rippleTouchController:insertRippleView:intoView: is not implemented, the rippleview will be
 automatically added as the top subview to the given view when the first tap event is happening.
 When @c rippleTouchController:insertRippleView:intoView: is implemented, it's the responsibility of
 the delegate to add the ripple view in the proper position within view's hierarchy if the delegate
 method is called.

 @param view The view that responds to the touch events for the ripple. The ripple
 is added to it as a subview.
 @param deferred Wheter the insertion of the rippleView to the provided view should be
 happened deferred
 @return an MDCRippleTouchController instance.
 */
```

romoore commented on components/Ripple/src/MDCRippleTouchController.m:
```diff
@@ -15,7 +15,28 @@
+  return self;
 }
 
 - (instancetype)initWithView:(UIView *)view {
```
Tip: Please have `initWithView:` call `initWithView:deferred:` instead of keeping the implementations independent.

romoore commented on components/Ripple/src/MDCRippleTouchController.m:
```diff
@@ -38,70 +59,107 @@ - (instancetype)init {
+  _delegate = delegate;
+
+  _delegateFlags.rippleTouchControllerShouldProcessRippleTouchesAtTouchLocation =
+      [_delegate respondsToSelector:@selector(rippleTouchController:
```
Tip: Since `respondsToSelector` is a runtime method, does it make sense to cache the values at assignment rather than calling it as-needed?  Maybe the caching behavior should be controlled via the new delay flag?

I know it's an edge case, but at a minimum it looks like the property should be updated to document that the delegate's behavior is cached at assignment rather than inspected at runtime. Since runtime inspection is typical for `@optional` protocol methods, I think it's helpful to point this out.  

For example, if the immediate delegate actually offloads delegate to other objects (it acts as a multiplexer), then it might not have configured all of its internal state when it's assigned as the delegate for this view. Later (before the view is in the view hierarchy) it has completed its own configuration (perhaps in `viewDidLoad` of a view controller) and will respond correctly.

maicki replied:
We are using this pattern across the Texture code base quit a bit: https://github.com/texturegroup/texture and I didn't hear any problems with it yet. I let it is for now, that said if @yarneo or @jverkoey think are for changing it, I'm happy to do this.

jverkoey replied:
I expect the edge case outlined here to be pretty rare, but in the interest of transparency and elevating this pattern a snippet of docs can go a long way. Consider adding something to the effect of:

> The delegate's behavior - in terms of which optional methods are deemed to be
> implemented - is cached at assignment rather than inspected on each invocation.

maicki replied:
Sounds great! Added.

romoore commented on components/Tabs/BUILD:
```diff
@@ -198,6 +198,7 @@ mdc_snapshot_objc_library(
         ":privateTabBarView",
         "//components/AppBar",
         "//components/HeaderStackView",
+        "//components/Ripple:private",
```
Can this be reverted now?

romoore commented on components/Ripple/BUILD:
```diff
@@ -27,6 +27,23 @@ load("@build_bazel_rules_swift//swift:swift.bzl", "swift_library")
+    deps = [":Ripple"],
+)
+
+package_group(
```
Can this file be reverted now?

romoore commented on components/Ripple/src/MDCRippleTouchController.m:
```diff
@@ -15,7 +15,28 @@
 #import "MDCRippleTouchController.h"
 
 @implementation MDCRippleTouchController {
+  MDCRippleView *_rippleView;
```
Nit: Why is this being redeclared as an explicit ivar rather than an `@synthesize`? Does it have side effects different from an iVar?  

I ask because in our code when I see an ivar directly allocated in the implementation, I interpret it as being fully-internal. When I see an `@synthesize` directive, I read it as storing the value for a public property.

romoore reviewed (commented):
I'm a little concerned at the idea of caching the selector response behavior since I've worked on code previously that would be incompatible with this approach.  That said, I know the behavior is not common and for most code caching makes sense.  I'm removing my "Request changes" since I don't know that this will break any clients.

Other than that, just some questions and suggestions sprinkled throughout.

maicki committed:
Code review changes

maicki commented:
@romoore Thank you for the review! I updated the PR based on it, please let me know in case I missed something.

cc @yarneo @jverkoey any other improvements to do or good to go?

maicki committed:
Add comment to MDCRippleTouchControllerDelegate caching
